README.md
[![codebeat badge][cq-image]][cq-url]
[![npm Version][npm-image]][npm-url]
[![License][license-image]][license-url]

# addMonths

Adds or sustract _X_ months to any JavaScript Date.

Please read the [What's New](#whats-new) section.

## Install

npm

```bash
npm i addmonths --save
```

bower

```bash
bower install addmonths --save
```

browsers

```html
<script src="https://unpkg.com/addmonths/dist/addmonths.js"></script>
<!-- ...or to inject `addMonths` to `Date.prototype` -->
<script src="https://unpkg.com/addmonths/dist/addmonths.auto.js"></script>
```

## Usage

```js
newdate = addMonths(startdate, count)
```

The `addMonths` function returns the date occurring _count_ months after _startdate_ or, if _count_ is negative, the date occurring _count_ months before _startdate_.

If _startdate_ is not a date, returns _startdate_ without changes.

If _count_ is falsy (i.e. zero), returns a new `Date` object with the same value as _startdate_.


### Examples

```js
const addMonths = require('addmonths')

addMonths(new Date(2017, 0, 1), 2)    // 2017-01-01 ==> 2017-03-01
addMonths(new Date(2017, 0, 1), -1)   // 2017-01-01 ==> 2016-12-01
addMonths(new Date(2016, 0, 31), 1)   // 2016-01-31 ==> 2016-02-29
addMonths(new Date(2016, 0, 31), 0)   // new instance with the same value
addMonths('2017-06-12', 1)            // the same string
```

and if you load addmonths/auto:

```js
require('addmonths/auto')

const date = new Date().addMonths(3)
```

## What's New

* Production ready version.
* Updated tests.
* Updated devDependencies.

## License

The [MIT License](LICENCE) (MIT)

Copyright (c) 2017 Alberto Martínez (https://github.com/aMarCruz)

[npm-image]:      https://img.shields.io/npm/v/addmonths.svg
[npm-url]:        https://www.npmjs.com/package/addmonths
[license-image]:  https://img.shields.io/npm/l/express.svg
[license-url]:    https://github.com/aMarCruz/jscc-brunch/blob/master/LICENSE
[cq-image]:       https://codebeat.co/badges/c44cf621-7412-4d0d-a132-cf09ee560d48
[cq-url]:         https://codebeat.co/projects/github-com-amarcruz-addmonths-master


CHANGES.md
# Changes for addMonths

#### oct/2017 - v1.0.0
* Production ready version.
* Updated tests.
* Updated devDependencies.

#### oct/2017 - v0.3.1
* Fixes typo in the README.

#### jul/2017 - v0.3.0

* Returns NaN for Dates evaluated to NaN.
* Include TypeScript def (CommonJS).
* Moved karma configuration to spec/.

#### jun/2017 - v0.2.0

Lots of changes in this version.

* You can use `addMonths` with UTC dates, avoiding call `addUTCMonths` or something (this is a not fully-tested feature).
* Now you can `require('addmonths/auto')` or load `dist/addmonths.auto.js` to inject an `addMonths` method into the `Date.prototype` that becomes the same as the function, but always returns a new instance of the date (and does not requires the first parameter of course).
* The filenames in the "dist" folder has changed. Now they are:
  - addmonths.js - [UMD](https://github.com/umdjs/umd) build for the browser (`window.addMonths`), AMD, and node or browserify-like loaders.
  - addmonths.esm.js - ES6 module (only `import` and `export`, the code is ES5).
  - addmonths.auto.js - Auto-running function (IIFE) to inject `Data.prototype.addMonths`.
* Refactorization of the tests, using [Karma](https://karma-runner.github.io/) and [Jasmine](https://jasmine.github.io/).
* Bower support.

#### jun/2017 - v0.1.0

First public release.


spec/index.js
'use strict'

/* eslint no-var:0 */

var addMonths = require('../')
var exitCode = 0

function dt(y, m, d, hr, min, sec, ms) {
  return new Date(y, m - 1, d, hr | 0, min | 0, sec | 0, ms | 0)
}

function err(data, result) {
  var cnt = (data[0] > 0 ? ' + ' : ' - ') + Math.abs(data[0])
  console.error(' Error with: ' + data[1].toLocaleString() + cnt)
  console.error('   expected: ' + data[2].toLocaleString())
  console.error('        got: ' + result.toLocaleString())
  console.error()
  exitCode = 1
}

var table = [
  [1, dt(2016, 1, 31), dt(2016, 2, 29)],      // 2016 was a leap year
  [1, dt(2017, 1, 31), dt(2017, 2, 28)],
  [1, dt(2018, 1, 1),  dt(2018, 2, 1)],
  [-1, dt(2018, 1, 1), dt(2017, 12, 1)],
  [12, dt(1900, 2, 3), dt(1901, 2, 3)]
]

console.log('Testing addMonths...')
for (var i = 0; i < table.length; i++) {
  var data = table[i]

  var result = addMonths(data[1], data[0])
  if (+result !== +data[2]) {
    err(data, result)
  }
}

if (exitCode) {
  process.exit(1)
} else {
  console.log('Done.')
}


spec/node.spec.js
const SpecReporter = require('jasmine-spec-reporter').SpecReporter

jasmine.getEnv().clearReporters()
jasmine.getEnv().addReporter(new SpecReporter({}))

global.addMonths = require('../')
require('../auto')
require('./test-info.js')
require('./test.spec.js')


spec/test-info.js
(function (_this) {

  _this.infoDates = {
    incX: [
      ['2017-01-01',  1, '2017-02-01'],
      ['2017-01-01', 12, '2018-01-01'],
      ['2017-01-01', 25, '2019-02-01'],
      ['9999-12-31', 12, '10000-12-31'],
      ['^10000,12,31,1,40,15', 2400, '^10200,12,31,1,40,15']
    ],
    decX: [
      ['2017-01-01',  -1, '2016-12-01'],
      ['2017-01-01', -12, '2016-01-01'],
      ['2017-01-01', -25, '2014-12-01']/*,
      ['Feb 1 100 12:00:00 GMT-0000',  -2400, '^-100,2,1,12']*/
    ],
    preserveTime: [
      ['2017-01-01 11:59:59.999', -1, '2016-12-01 11:59:59.999'],
      ['Jan 7 2017 23:59:00 GMT-0500', -12, 'Jan 7 2016 23:59:00 GMT-0500'],
      ['Jan 7 2017 23:59:00 GMT-0500', -25, 'Dec 7 2014 23:59:00 GMT-0500'],
      ['Jan 7 2017 23:59:00 GMT-1100', -25, 'Dec 7 2014 23:59:00 GMT-1100']
    ],
    limitToLastDay: [
      ['2017-01-31',  1, '2017-02-28'],
      ['2016-10-31',  1, '2016-11-30'],
      ['2018-03-31', -1, '2018-02-28'],
      ['2018-07-31', -1, '2018-06-30']
    ],
    leapYears: [
      ['2016-01-31',  1, '2016-02-29'],
      ['2016-02-29', 12, '2017-02-28'],
      ['2016-01-31 12:00:00',  1, '2016-02-29 12:00:00'],
      ['2016-01-31 11:59:59',  1, '2016-02-29 11:59:59'],
      ['2016-01-31 00:00:00',  1, '2016-02-29 00:00:00'],
      ['2021-03-31 00:00:00', -1, '2021-02-28 00:00:00'],
      ['2021-01-31T00:00:00Z', 1, '2021-02-28T00:00:00Z']
    ]
  }

  // for browsers without support for non-common dates (PhantomJS)
  if (isNaN(new Date(10200, 11, 31))) {
    _this.infoDates.incX.pop()
    _this.infoDates.decX.pop()
  }

})(typeof window == 'object' ? window : global)


src/addmonths.js

/**
 * Adds or subtract X months to any date.
 * Use a negative number to subtract.
 * If there is an overflow in the day, the date is adjusted to the last valid day of the month.
 *
 * This function does not change the original date.
 *
 * @param   {Date}   startdate  - A JavaScript Date object
 * @param   {number} count      - Number of months to add or substract
 * @returns {Date}   New, adjusted date.
 */
export default function addMonths(startdate, count) {

  // Invalid dates created with the Date() constructor have getDate
  // method so we will to re-check with the copy of startdate.
  if (startdate && startdate.getDate) {

    // Make a copy
    const date = new Date(+startdate)

    // Skip work if date is invalid or there's nothing to change.
    if (!isNaN(date) && (count |= 0)) {

      // The trick here is set the time to midday, so timezone does
      // not change the day or month of the resulting date.
      const offset = +startdate - date.setUTCHours(12, 0, 0, 0)
      const day = date.getUTCDate()

      // Set the new month with day 2 to make sure the month does not overflow
      date.setUTCMonth(date.getUTCMonth() + count, 2)

      // ...and grab the resulting (and correct) month
      const month = date.getUTCMonth()

      // Now, restore day and time
      date.setUTCDate(day)

      // If the month has changed, we have an overflow, so set
      // the date to the last day of the previous month.
      if (date.getUTCMonth() !== month) {
        date.setUTCDate(0)
      }

      // We have the correct day and month, now restore the time offset.
      date.setMilliseconds(offset)

    }

    return date
  }

  return startdate
}


karma.conf.js
// Karma configuration
// Generated on Mon Jun 12 2017 20:43:23 GMT-0500 (CDT)

/*
const saucelabsBrowsers = require('./spec/saucelabs-browsers').browsers
const isSaucelabs = process.env.SAUCELABS

// run the tests only on the saucelabs browsers
const browsers = isSaucelabs
    ? Object.keys(saucelabsBrowsers) : ['PhantomJS', 'Chrome']*/
const browsers = ['PhantomJS', 'Chrome', 'Firefox']

module.exports = function (config) {
  config.set({

    basePath: '',

    frameworks: ['jasmine'],

    files: [
      'dist/addmonths.js',
      'dist/addmonths.auto.js',
      'spec/test-info.js',
      'spec/test.spec.js'
    ],

    exclude: [],

    preprocessors: {},

    reporters: ['progress'],

    port: 9876,

    colors: true,

    // config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_WARN,

    autoWatch: true,

    browsers: browsers,

    // if true, Karma captures browsers, runs the tests and exits (for CI)
    singleRun: true,

    concurrency: Infinity
  })
}
